core/lib/contracts/src/consensus.rs

```diff
@@ -0,0 +1,323 @@
+/// Strongly-typed API for Consensus-related solidity contracts.
+/// Technically this code could be autogenerated from the json spec of the contract,
+/// but we don't use any such autogen tooling currently.
+use std::sync::Arc;
+
+use anyhow::Context as _;
+use ethabi::Token;
+
+// First we define some traits that the types representing concrete contracts
+// will implement.
+
+/// Strongly typed representation of a contract function.
+/// It also represents the inputs of the function.
+pub trait Function {
+    /// Name of the solidity function.
+    const NAME: &'static str;
+    /// Type representing contract this function belongs to.
+    type Contract: AsRef<ethabi::Contract>;
+    /// Typ representing outputs of this function.
+    type Outputs;
+    /// Encodes this struct to inputs of this function.
+    fn encode(&self) -> Vec<Token>;
+    /// Decodes outputs of this function.
+    fn decode_outputs(outputs: Vec<Token>) -> anyhow::Result<Self::Outputs>;
+}
+
+/// Address of contract C. It is just a wrapper of ethabi::Address,
+/// just additionally indicating what contract is deployed under this address.
+#[derive(Debug)]
+pub struct Address<C>(ethabi::Address, std::marker::PhantomData<C>);
+
+impl<C> Clone for Address<C> {
+    fn clone(&self) -> Self {
+        *self
+    }
+}
+
+impl<C> Copy for Address<C> {}
+
+impl<C> PartialEq for Address<C> {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.eq(&other.0)
+    }
+}
+
+impl<C> Eq for Address<C> {}
+
+impl<C> Address<C> {
+    pub fn new(address: ethabi::Address) -> Self {
+        Self(address, std::marker::PhantomData)
+    }
+}
+
+impl<C> std::ops::Deref for Address<C> {
+    type Target = ethabi::Address;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+/// Represents a call to the function F.
+pub struct Call<F: Function> {
```

Comment:
I don't want to give up on the strongly typed contract interfaces. So you want me to move all this stuff to consensus crate?

Reply:
But it is strongly typed (well, it _can_ be), `CallFunctionArgs::new` accept `impl Tokenize`, and the call will use `Detokinize` for result, so you can define inputs/outputs and implement corresponding traits for them. Nothing also prevents you from creating a struct like `AttesterCommittee` which will have method that will hold the contract ABI and will have concrete methods that do `ContractCall::new(..).for_contract(..).call(..).await` under the hood. Moreover, likely you will want to implement (De)Tokenize later anyway, at least for some of the types, -- I guess `EthWatch` will want to listen for some consensus-related events on the contract.

If you want to keep it, I think at least for now it's indeed better to put it into the consensus crate and expand the visibility on demand.

Reply:
What eth_client does is not static typing: instead of accepting/returning a sequence of tokens, it accepts whatever converts to a sequence of tokens. This is just a syntax sugar.
This abstraction is basically a placeholder until I can use https://docs.rs/alloy-sol-types/

The eth_client api doesn't provide means to express static relation between
- concrete contract and its address
- concrete function and the concrete contract is belongs to
- concrete function and its input/output types
and it lacks means to write test the encoding/decoding against the contract schema.

It is likely possible to make consensus::vm::VM implement EthInterface, but it is entirely orthogonal to static typing.

Reply:
> This abstraction is basically a placeholder

That's my main concern. By the time we will get there, we will have places that use `CallArguments`, places that use `Tokenize/Detokenize`, places that use `ethabi::Token` directly, and places that _(new new)_ use `zksync_contract::Call`. Given that we're, ugh, not always prioritizing the tech debt, I'd expect it to be messy for a while. So I want to minimize the surface of changes, once we're ready for them.

That said, I see your point (as I mentioned above, I'm not requesting changes -- it was a genuine question). I still however think that it'd be better to put this code to the consensus crate until we need it elsewhere (and once we need it elsewhere, discuss whether a better idea would be to resource the migration to `alloy`).

Reply:
moved to consensus crate